src/sql/statement.rs

```diff
@@ -1169,7 +1185,48 @@ fn handle_create_source(scx: &StatementContext, stmt: Statement) -> Result<Plan,
                 }
             };
 
-            let mut desc = encoding.desc(envelope)?;
+            let envelope = match &envelope {
+                sql_parser::ast::Envelope::None => dataflow_types::Envelope::None,
+                sql_parser::ast::Envelope::Debezium => dataflow_types::Envelope::Debezium,
+                sql_parser::ast::Envelope::Upsert(key_format) => match connector {
+                    Connector::Kafka { .. } => {
+                        let mut key_encoding = if key_format.is_some() {
+                            get_encoding(key_format, std::collections::HashMap::new())?
+                        } else {
+                            encoding.clone()
+                        };
+                        // Shift the key_schema to the value_schema position
+                        // for validation.
+                        if let DataEncoding::Avro(AvroEncoding {
+                            key_schema,
+                            value_schema,
+                            ..
+                        }) = &mut key_encoding
```

Comment:
It seems to me that `key_schema` should be removed from `AvroEncoding`, so that we can have two separate `DataEncoding`s: one for the key and one for the value. With `Envelope::None` or `Envelope::Debezium, the key encoding is used only to determine what the key columns are, but with `Envelope::Upsert` the key encoding is used to actually inject the key columns into the data plane.

It's also possible that it's only `Envelope::Debezium` where the key is duplicated in the value, and `Envelope::None` should behave more similarly to `Envelope::Upsert` and inject the key into the data plane.

Hrmmmm. @umanwizard do you have thoughts here?